langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java

```diff
@@ -0,0 +1,693 @@
+        ensureNotNull(request, "request");
+
+        SQLFilter sqlFilter = SQLFilters.create(request.filter(), metadataKeyMapper);
+        final int maxResults = request.maxResults();
+
+        // In a 23.4 build of Oracle Database, ORA-06553 will result if the distance column is referenced in the WHERE
+        // clause. For this reason, the minScore filtering will happen locally. There are workarounds, such as binding
+        // the VECTOR twice, or using a sub-select. These can be implemented if proven to offer a significant
+        // performance improvement.
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement query = connection.prepareStatement(
+                     "SELECT VECTOR_DISTANCE(" +
+                             table.embeddingColumn() + ", ?, " + distanceMetric.name() + ") distance, " +
+                             String.join(", ", table.idColumn(), table.embeddingColumn(), table.textColumn(),
+                                     table.metadataColumn()) +
+                         " FROM " + table.name() +
+                         sqlFilter.asWhereClause() +
+                         " ORDER BY distance" +
+                         " FETCH " + (isExactSearch ? "" : " APPROXIMATE") +
+                         " FIRST " + maxResults + " ROWS ONLY")
+        ) {
+            query.setObject(1, request.queryEmbedding().vector(), OracleTypes.VECTOR_FLOAT32);
+            sqlFilter.setParameters(query, 2);
+            query.setFetchSize(maxResults);
+
+            // Calls to defineColumnType reduce the number of network requests. When Oracle JDBC knows that it is
+            // fetching VECTOR, CLOB, and/or JSON columns, the first request it sends to the database can include a LOB
+            // prefetch size (VECTOR and JSON are value-based-lobs). If defineColumnType is not called, then JDBC needs
+            // to send an additional request with the LOB prefetch size, after the first request has the database
+            // respond with the column data types.
+            OracleStatement oracleStatement = query.unwrap(OracleStatement.class);
+            oracleStatement.defineColumnType(1, OracleTypes.BINARY_DOUBLE);
+            oracleStatement.defineColumnType(2, OracleTypes.VARCHAR);
+            oracleStatement.defineColumnType(3, OracleTypes.VECTOR_FLOAT32, 524308); // <-- Max vector size, in bytes
+            oracleStatement.defineColumnType(4, OracleTypes.CLOB, Integer.MAX_VALUE);
+            oracleStatement.defineColumnType(5, OracleTypes.JSON, Integer.MAX_VALUE);
+
+            List<EmbeddingMatch<TextSegment>> matches = new ArrayList<>(maxResults);
+            try (ResultSet resultSet = query.executeQuery()) {
+                while (resultSet.next()) {
+
+                    // Convert the cosine distance between 0 and 2 into a score between 1 and 0.
+                    double score = 1d - (resultSet.getDouble("distance") / 2d);
```

Comment:
It probably makes sense to do this conversion only when `COSINE` distance metric is used

Reply:
I'm very glad you noticed this, because I completely missed it!

I'd like to get your opinion: Do you think that supporting distance metrics, other than cosine, makes sense? I ask this question in the context of langchain4j, where the EmbeddingMatch.score() method is specified as:
```java
    /**
     * Returns the relevance score (derivative of cosine distance) of this embedding compared to
     * a reference embedding during a search.
     * The current implementation assumes that the embedding store uses cosine distance when comparing embeddings.
     *
     * @return Relevance score, ranging from 0 (not relevant) to 1 (highly relevant).
     */
```
Based on this, it seems that an implementation of `search(EmbeddingRequest)` is only correct if it uses cosine distance. So, if we allow users to configure a different metric, then we are allowing them to create an instance of EmbeddingStore which does not behave correctly.

Do you think we should remove the configurable distance metric, and just to stick to cosine?

Reply:
While cosine is probably the most used one, current API is overly strict in this regard and should be changed to support more metrics.
It probably makes little sense to support other metrics till then. Or at least invert them to make sure the returned score and order are correct. I am fine with both options

Reply:
I think the best move is to remove support for non-cosine metrics for now. It keeps the user experience simpler, and it keeps our code simpler. Once the API is ready, we can add the support back, and we can do it in a way that is consistent with all other embedding stores.

Reply:
Sounds good!